pkg/proxy/iptables/proxier.go

```diff
@@ -391,16 +297,19 @@ func NewProxier(ipt utiliptables.Interface,
 
 	if len(clusterCIDR) == 0 {
 		glog.Warningf("clusterCIDR not specified, unable to distinguish between internal and external traffic")
+	} else if utilproxy.IsIPv6CIDR(clusterCIDR) != ipt.IsIpv6() {
+		return nil, fmt.Errorf("clusterCIDR %s has incorrect IP version: expect isIPv6=%t", clusterCIDR, ipt.IsIpv6())
 	}
 
 	healthChecker := healthcheck.NewServer(hostname, recorder, nil, nil) // use default implementations of deps
 
+	isIPv6 := ipt.IsIpv6()
 	proxier := &Proxier{
 		portsMap:                 make(map[utilproxy.LocalPort]utilproxy.Closeable),
 		serviceMap:               make(proxy.ServiceMap),
-		serviceChanges:           proxy.NewServiceChangeTracker(),
+		serviceChanges:           proxy.NewServiceChangeTracker(customizeServiceInfo, &isIPv6, recorder),
```

Comment:
I feel like we should not enforce the construction of ServiceInfo at ServiceChangeTracker level. It should be done at each Update. Let us say ServiceInfo is very complicated and there is missing information at init time of ServiceChangeTracker.

The same for EndpointChangeTracker

Reply:
Sorry. I take it back. Let me think if I can come up with a better way to structure this.